src/tdigest_api.c

```diff
@@ -0,0 +1,693 @@
+
+	// compress if beyond threshold
+	if (list_length(tts->unmerged_centroids) > tts->td->threshold)
+		ts_tdigest_trans_compress(tts);
+}
+
+/*
+ * Transition function for TDigest aggregate. Starts up aggregate if this is first value being
+ * added.
+ * Args: TDigestTState, Value to add
+ */
+TS_FUNCTION_INFO_V1(ts_tdigest_trans_sql);
+Datum
+ts_tdigest_trans_sql(PG_FUNCTION_ARGS)
+{
+	MemoryContext old, context;
+	TDigestTState *tts;
+
+	if (!AggCheckCallContext(fcinfo, &context))
+		elog(ERROR, "ts_tdigest_trans_sql called in non-aggregate context");
+
+	old = MemoryContextSwitchTo(context);
+
+	if (PG_ARGISNULL(0))
+		tts = ts_tdigest_startup(0);
+	else
+		tts = (TDigestTState *) PG_GETARG_VARLENA_P(0);
+
+	// add data to TDigest
+	if (!PG_ARGISNULL(1))
+		ts_tdigest_add(tts, PG_GETARG_FLOAT8(1), 1);
+
+	MemoryContextSwitchTo(old);
+
+	PG_RETURN_POINTER(tts);
+}
+
+/*
+ * Transition function for TDigest aggregate. Starts up aggregate with provided compression
+ * parameter if this is first value being added.
+ * Args: TDigestTState, Value to add, Compression
+ */
+TS_FUNCTION_INFO_V1(ts_tdigest_trans_comp_sql);
+Datum
+ts_tdigest_trans_comp_sql(PG_FUNCTION_ARGS)
+{
+	MemoryContext old, context;
+	TDigestTState *tts;
+	int32_t compression = PG_GETARG_INT32(2);
+
+	if (!AggCheckCallContext(fcinfo, &context))
+		elog(ERROR, "ts_tdigest_trans_comp_sql called in non-aggregate context");
+
+	old = MemoryContextSwitchTo(context);
+
+	// set up or retrieve tdigest
+	if (PG_ARGISNULL(0))
+		tts = ts_tdigest_startup(compression);
+	else
+		tts = (TDigestTState *) PG_GETARG_VARLENA_P(0);
+
+	// add element to a tdigest
+	if (!PG_ARGISNULL(1))
+		ts_tdigest_add(tts, PG_GETARG_FLOAT8(1), 1);
+
+	MemoryContextSwitchTo(old);
+
+	PG_RETURN_POINTER(tts);
+}
+
+/*
+ * Merges two TDigest aggregate transition states
+ * Data from second arg is merged into first arg
+ */
+static void
+ts_tdigest_trans_merge(TDigestTState *tts1, TDigestTState *tts2)
+{
+	int i;
+	TDigest *t1, *t2;
+
+	t1 = tts1->td;
+
+	// compress before fetching data to merge in
+	ts_tdigest_trans_compress(tts2);
+	t2 = tts2->td;
+
+	for (i = 0; i < t2->num_centroids; i++)
+	{
+		Centroid *c = &t2->centroids[i];
+		ts_tdigest_add(tts1, c->mean, c->weight);
+	}
+}
+
+/*
+ * Merges two TDigest aggregates.
+ * Args: TDigestTState, TDigestTState
+ */
+TS_FUNCTION_INFO_V1(ts_tdigest_combine_sql);
+Datum
+ts_tdigest_combine_sql(PG_FUNCTION_ARGS)
+{
+	MemoryContext old, context;
+	TDigestTState *receiver_tts, *incoming_tts;
+
+	if (!AggCheckCallContext(fcinfo, &context))
+		elog(ERROR, "dist_combine called in non-aggregate context");
+
+	if (PG_ARGISNULL(0) && PG_ARGISNULL(1))
+		PG_RETURN_NULL();
+
+	old = MemoryContextSwitchTo(context);
+
+	if (PG_ARGISNULL(0)) // compress unmerged and reallocate in correct memory context
+	{
+		incoming_tts = (TDigestTState *) PG_GETARG_VARLENA_P(1);
+		ts_tdigest_trans_compress(incoming_tts);
+
+		receiver_tts = ts_tdigest_startup(0);
```

Comment:
ts_tdigest_startup allocs and inits td. We are overwriting that here. Saw this in a few other places too. Need a version of startup like ts_tdigest_alloc_and_set( td) that takes in the passed in td and uses it.